File: document/JAVA/sort/python/quick_sort.py
"""
Quick Sort (Lomuto partition)
- Stable: No
- In-place: Yes
- Time: O(n log n) average, O(n^2) worst (e.g., sorted input with poor pivot)
- Space: O(log n) recursion average
"""

def quick_sort(arr):
    def partition(a, low, high):
        pivot = a[high]
        i = low
        for j in range(low, high):
            if a[j] <= pivot:
                a[i], a[j] = a[j], a[i]
                i += 1
        a[i], a[high] = a[high], a[i]
        return i

    def qs(a, low, high):
        if low < high:
            p = partition(a, low, high)
            qs(a, low, p - 1)
            qs(a, p + 1, high)

    qs(arr, 0, len(arr) - 1)
    return arr

if __name__ == "__main__":
    data = [10, 7, 8, 9, 1, 5]
    print("before:", data)
    print("after :", quick_sort(data))

File: document/JAVA/sort/c/quick_sort.c
#include <stdio.h>
#include <stdbool.h>

// Utility: print array
static void print_array(int *a, int n) {
    for (int i = 0; i < n; ++i) {
        if (i) printf(" ");
        printf("%d", a[i]);
    }
    printf("
");
}

/*
Quick Sort (Lomuto)
- Stable: No
- In-place: Yes
- Time: O(n log n) avg, O(n^2) worst
- Space: O(log n) recursion avg
*/
static int partition(int *a, int low, int high) {
    int pivot = a[high];
    int i = low;
    for (int j = low; j < high; ++j) {
        if (a[j] <= pivot) {
            int t = a[i]; a[i] = a[j]; a[j] = t;
            ++i;
        }
    }
    int t = a[i]; a[i] = a[high]; a[high] = t;
    return i;
}

static void qsort_rec(int *a, int low, int high) {
    if (low < high) {
        int p = partition(a, low, high);
        qsort_rec(a, low, p - 1);
        qsort_rec(a, p + 1, high);
    }
}

int main(void) {
    int a[] = {10,7,8,9,1,5};
    int n = sizeof(a)/sizeof(a[0]);
    qsort_rec(a, 0, n - 1);
    print_array(a, n);
    return 0;
}
